typed/writer.go

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package typed
+
+import (
+	"encoding/binary"
+	"io"
+	"sync"
+)
+
+type intBuffer [8]byte
+
+var intBufferPool = sync.Pool{New: func() interface{} {
+	return new(intBuffer)
+}}
+
+// Writer is a writer that writes typed values to an io.Writer
+type Writer struct {
+	writer io.Writer
+	err    error
+}
+
+// NewWriter creates a writer that writes typed value to a reader
+func NewWriter(w io.Writer) *Writer {
+	return &Writer{
+		writer: w,
+	}
+}
+
+// WriteBytes writes a slice of bytes to the io.Writer
+func (w *Writer) WriteBytes(b []byte) {
+	if w.err != nil {
+		return
+	}
+
+	if _, err := w.writer.Write(b); err != nil {
+		w.err = err
+	}
+}
+
+// WriteUint16 writes a uint16 to the io.Writer
+func (w *Writer) WriteUint16(n uint16) {
+	if w.err != nil {
+		return
+	}
+
+	sizeBuf := intBufferPool.Get().(*intBuffer)
+	defer intBufferPool.Put(sizeBuf)
+
+	binary.BigEndian.PutUint16(sizeBuf[:2], n)
```

Comment:
I don't have the context of why we used sync pool here, but would it be faster than a stack variable here?

like `var [2]byte`

Reply:
passing a slice pointing to a `[2]byte` into `io.Writer.Write()` causes that block of memory to escape from the function, resulting in an alloc.

I verified it via:
```golang
package tchannel

import (
	"io/ioutil"
	"sync"
	"testing"
)

var pool = sync.Pool{New: func() interface{} {
	return new([8]byte)
}}

func fooWithPool() error {
	sizeBuf := pool.Get().(*[8]byte)
	defer pool.Put(sizeBuf)

	if _, err := ioutil.Discard.Write(sizeBuf[:]); err != nil {
		return err
	}

	return nil
}

func foo() error {
	var sizeBuf [8]byte

	if _, err := ioutil.Discard.Write(sizeBuf[:]); err != nil {
		return err
	}

	return nil
}

func BenchmarkArrayWithPool(b *testing.B) {
	var err error
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		if err = fooWithPool(); err != nil {
			break
		}
	}
}

func BenchmarkArray(b *testing.B) {
	var err error
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		if err = foo(); err != nil {
			break
		}
	}
}
```
which gives
```
$ go test -run=NONE -bench=BenchmarkArray -benchmem
goos: darwin
goarch: amd64
pkg: github.com/uber/tchannel-go
BenchmarkArrayWithPool-8   	57540184	        21.0 ns/op	       0 B/op	       0 allocs/op
BenchmarkArray-8           	74787500	        14.8 ns/op	       8 B/op	       1 allocs/op
PASS
ok  	github.com/uber/tchannel-go	2.747s
```